### DISCUSSION TITLE
Melhorias ao código

### DISCUSSION THREAD

**Discussion by FelipeSBarros** [FelipeSBarros/crossfire#1] [Q&A]
Este Python package foi criado considerando a versão 1 da API do [Fogo Cruzado](fogocruzado.org.br). Se trata de uma API para acessar aos dados gerados por [crowd sourcing](https://en.wikipedia.org/wiki/Crowdsourcing) sobre tiroteiros em algumas cidades do Brasil.
Recentemente, foi lançada a [versão 2](https://api.fogocruzado.org.br/docs) da API e gostaria de atualizar o módulo.
Com isso surgiram dúvidas:
- [x] Ao clonar o projeto e usar o poetry para instalar as dependencias, tive problema. [No último commmit](https://github.com/FelipeSBarros/crossfire/commit/c9a6a3a22029eca3c855791a3fa9844aa760fa65), acho que foi resolvido. Preciso confirmar;

**Sobre organização de código:**
- [ ] Inicialmente eu havia separado:
    - [fogocruzado_signin.py](https://github.com/FelipeSBarros/crossfire/blob/master/Python/crossfire/crossfire/fogocruzado_signin.py): Possui apenas uma função (`fogocruzado_signin`) que transforma os parâmetros (usuário e senha) em variáveis de ambiente (que serão usadas em outras funções. Mas reparem que não faz o `signin`, de fato. Começo a suspeitar que além do nome não corresponder com o resultado da função, essa função pode estar em `fogocruzado_utils.py`.
    - [fogocruzado_utils.py](https://github.com/FelipeSBarros/crossfire/blob/master/Python/crossfire/crossfire/fogocruzado_utils.py): Já esse arquivo possui cinco funções:
        - `fogocruzado_key`, que acessa e retorna a chave de acesso à API obtida pela função `get_token_fogocruzado`
        - `get_token_fogocruzado` acessa as variáveis de ambiente geradas no `fogocruzado_signin` e, em sendo chaves válidas, salva como variáveis de ambiente (a mesma resgatada pelo `fogocruzado_key`);
        - `extract_data_api` usa o `fogocruzado_key` para acessar o token e [endpoint](https://api.fogocruzado.org.br/docs/endpoint/occurrences) de ocorrencias de tiroteio. Reparem que ele recebe um parâmetro `link`. Isso pelo fato dele ser usado numa função "maior", em `get_fogocruzado.py`. Falo mais dele logo mais....
        - `get_cities_api` usar o `fogocruzado_key` para acessar o token e resgatar dados do [endpoint cities](https://api.fogocruzado.org.br/). Os dados retornados são convertidos em `DataFrame`.
        - `get_states_api` similar ao  `get_cities_api`, mas dessa vez acessa o [endpoint de estados](https://api.fogocruzado.org.br/docs/endpoint/states). Os dados retornados são convertidos em `DataFrame`;

**Sobre essas cinco funções, tenho algumas dúvidas:**
* Aqueles que acessam a API, poderiam ter a URL como uma constante, assim, cada função de acesso aos dados poderia usa-la e alterar o final da url, conforme necessário. **Faz sentido?** Alguma sugestão de módulo python para manipulação de `URL`?
* Ao acessar o endpoint da API de cidades, [pode-se apresentar alguns parâmetros de filtro](https://api.fogocruzado.org.br/docs/endpoint/cities#exemplos-de-uso). Qual seria a melhor abordagem para incorporar essa possibilidade no código?
    * Haveria alguma "boa prática" ao considerar que os parâmetro podem vir do resultado do endpoint de *states*?
    
Continuando....
    - [get_cities.py](https://github.com/FelipeSBarros/crossfire/blob/master/Python/crossfire/crossfire/get_cities.py): É um script usa o `get_citites_api` e  apenas convertir parte dos dados. Contudo, pelo que vi na documentação, esse dado em questão já não é mais retornado pelo endpoint. Ou sjea, não tem mais motivo de exitir.
    - e [get_fogocruzado.py](https://github.com/FelipeSBarros/crossfire/blob/master/Python/crossfire/crossfire/get_fogocruzado.py): essa é a função usada para estruturar a URL do [`endpoint` de ocorrencias](https://api.fogocruzado.org.br/docs/endpoint/occurrences) e retornorará as ocorrencias de tiroteio em formato [`GeoDataFrame`](https://geopandas.org/en/stable/). É por ela que fica 'a logica de negócio' dos filtros... o que está, não necessáriamente permanecerá. Pois entendo que algumas restrições serão desconsideradas. [Mas reparem que há a possibilidade de filtro por data(s) e cidade(s)](https://api.fogocruzado.org.br/docs/endpoint/occurrences#exemplos-de-uso).

**Demais Dúvidas**
Daí surge a pergunta, que em muito se assemelha à pergunta anterior: como fazer com que o filtro de cidade, possa receber o resultado de um retorno à endpoint de cidades da API E também a uma consulta por nome de cidade?
Faz sentido separar tanto o código, também?

**Comment by FelipeSBarros** [comment]
A correção do poetry funcionou bem:
```
Installing dependencies from lock file

Package operations: 21 installs, 1 update, 0 removals

  • Installing click (8.1.7)
  • Installing six (1.16.0)
  • Installing attrs (23.1.0)
  • Installing certifi (2023.7.22)
  • Installing click-plugins (1.1.1)
  • Installing cligj (0.7.2)
  • Installing numpy (1.25.2)
  • Installing python-dateutil (2.8.2)
  • Installing pytz (2023.3)
  • Updating setuptools (65.6.3 -> 68.1.2)
  • Installing charset-normalizer (3.2.0)
  • Installing fiona (1.9.4.post1)
  • Installing idna (3.4)
  • Installing pandas (1.5.3)
  • Installing pyproj (3.6.0)
  • Installing shapely (2.0.1)
  • Installing urllib3 (2.0.4)
  • Installing zope-interface (6.0)
  • Installing datetime (4.9)
  • Installing geopandas (0.10.2)
  • Installing python-decouple (3.8)
  • Installing requests (2.31.0)

Installing the current project: crossfire (0.2.0.2)
```

**Comment by cuducos** [comment]
Acho que o ponto principal do pacote não é a manipulação das variáveis de ambiente, mas [obter o token](https://github.com/FelipeSBarros/crossfire/blob/c9a6a3a22029eca3c855791a3fa9844aa760fa65/Python/crossfire/crossfire/fogocruzado_signin.py#L24C5-L24C26) utilizado para authenticação.

Acredito que o melhor seja separar essas responsabilidades:
- ler variáveis de ambiente parte configuração, e poderia estar num módulo como `config.py` ou `settings.py`
- setar variáveis de ambiente é uma abstração muito baixa, poderíamos substituir por ler esse módulo de configuração (ou que eles seja uma classe e criemos instâncias dele)
- a parte de obtenção do token me parece algo para um serviço de `auth` que pode gerenciar coisas como obtenção, validação da expiração (caso tenha), e renovação do token (caso necessário)

Por exemplo:

```python
from decouple import config

# settings.py
class Settings:
    def __init__(self, **kwargs):
        self.email =  kwargs.get('email', config('FOGOCRUZADO_EMAIL'))

# auth
class Auth:
    def __init__(self, settings: Settings):
        self.settings = settings
     
     @property
     def token(self):
         # TODO: usar settings.email e settings.pasword para obter o token
         ....

# service.py (não lembro onde a gente usa isso, então pseudo-código mais do qq outra coisa)

settings = Settings()
auth = Auth(settings)
faz_requisicao(URL, token=auth.token)
```

Essa lógica do final pode ser abstraída numa classe `Client` com os recursos que vamos consultar.

Aí teríamos basicamente, três módulos principais:
1. Configuração: lê variáveis de ambiente, valida, e trata (por exemplo, converte `'42'` para `42`) as configurações da biblioteca
2. Autenticação: usa o objeto de configuração para obter as credenciais do servidor do Fogo Cruzado
3. Cliente: utiliza as credenciais da autenticação para abstrair as requisições HTTP, converter os dados (por exemplo, de texto para `GeoDataFrame`)

Acho que assim podemos matar o `fogocruzado_signin.py` e simplificar o `fogocruzado_utils.py`.

Sobre o restante, acho que o `get_cities.py` (pelo o que entendi, não é utilizado mais) pode sumir tb, e parece OK tuas considerações sobre o `get_fogocruzado.py`.

Eu só faria mais uma adição: pq numa biblioteca chamada `fogocruzado` (em inglês) tudo tem que ter o prefixo ou sufixo `fogocruzado`? Não vejo sentido. Só deixa o código mais verboso e não acrescenta nada.

Já que vamos fazer um [_major bump_](https://semver.org/), vamos renomear as coisas para evitar essa redundância?

**Comment by cuducos** [comment]
Pensando um pouco mais, acho que já que vamos de _major bump_, dá para ser mais arrojadas. Vamos reescrever as instruções do `README.md` pensando como seria o caso de uso ideal. Pode ser aqui numa discussão, ou numa _issue_, assim vamos coletando opiniões e aprimorando uma primeira versão. [Depois a gente escreve ou refatora o código para fazer essa experiência de uso acontecer](https://youtu.be/JQBnopNPQEc).

Por exemplo, acho isso ruim (e é o que está documentado hoje):

```python
>>> from crossfire import fogocruzado_signin  # desnecessário importar fogo cruzado de fogo cruzado
>>> fogocruzado_signin('user@host.com', 'password') # prática não recomendável colocar credenciais no código https://12factor.net/config
>>> from crossfire import get_fogocruzado  #  mesma coisa da 1ª linha… e meio desnecessário ter que utilizar dois imports
>>> fogocruzado = get_fogocruzado(state=['RJ'])  # nomes muito genéricos (get e nome do pacote)
```

Eu sugeriria algo assim

```python
>>> from crossfire import Client  # nomes não repetitivos, um único import
>>> client = Client()  # na hora de instanciar, podemos passar algum parâmetro CASO não queira usar os valores padrão (por exemplo, credenciais já configuradas nas variáveis de ambiente)
>>> data = client.occurrences(states='RJ')  # método mais preciso no nome, e tipagem dinâmica (internamente a gente converte de 'RJ' para ['RJ'])
```

Isso escrevo muito pensando [em ir além da PEP8](https://youtu.be/wf-BqAjZb8M).

Aí a gente documenta os casos personalizados… por exemplo, senão for de configurar `FOGOCRUZADO_EMAIL` e `FOGOCRUZADO_PASSWORD`:

```python
>>> from crossfire import Client, Settings
>>> settings = Settings(
...     email=os.envvar['OUTRO_EMAIL'], 
...     password=os.envvar['OUTRA_SENHA'],
...  )
>>> client = Client(settings)
```

E aí saímos escrevendo testes, refatorando e coisa e tal.

**Comment by FelipeSBarros** [comment]
Lendo com mais calma, fico com algumas dúvias:

> setar variáveis de ambiente é uma abstração muito baixa, poderíamos substituir por ler esse módulo de configuração (ou que eles seja uma classe e criemos instâncias dele)

Qual seira a diferença entre essas opções? Imagino que a primeira demandaria do usuário adicionar suas credenciais e algum arquivo que seria lido ao carregar o módulo e que seria usado para acessar o token. 
Já o segundo, seria a criação de uma clase com métodos de `login` como vc sugeriu no pseudocódigo...

É isso?

> a parte de obtenção do token me parece algo para um serviço de auth que pode gerenciar coisas como obtenção, validação da expiração (caso tenha), e renovação do token (caso necessário)
Entendi. Não sei como é a dinâmica do token... se ele precisa ser renovado. Vou perguntar ao Davi;

Achei legal essa abordagem!
```
def __init__(self, **kwargs):
        self.email =  kwargs.get('email', config('FOGOCRUZADO_EMAIL'))
```
Aliás, acabei de aprender que com essa abordagem, podemos deixar as variaveis de ambiente em um .env, podemos mandar por um kwargs e ainda como parẫmetro da função. Legal!

> Essa lógica do final pode ser abstraída numa classe `Client` com os recursos que vamos consultar.
Acho uma boa ideia. Só fico pensando, se poderiamos encontrar um nome mais legal... hahaha se depender de mim, seria fogocruzado. Mas sei, a idéia é que o nome represente a funcionalidade de sua exitencia, aém do fato de fogocruzado aumentar a redundância...

> Aí teríamos basicamente, três módulos principais:
> Configuração: lê variáveis de ambiente, valida, e trata (por exemplo, converte '42' para 42) as configurações da biblioteca
Ok, entendo que essa seria a `settings` proposta anteriomente (mais ou menos)...
> Autenticação: usa o objeto de configuração para obter as credenciais do servidor do Fogo Cruzado
OK. Já mandei a pergunta ao Davi sobre a "vida útil" do token... dai vemos se é o caso de ter métodos para atualização e etc...
> Cliente: utiliza as credenciais da autenticação para abstrair as requisições HTTP, converter os dados (por exemplo, de texto para GeoDataFrame)
Legal. Acho que é apresentada na sua ultima mensagem...

Em resumo, concordo com a abordagem. Nao teria qualquer sugestão de melhorias a fazer...

**Comment by cuducos** [comment]
*Configurar variável de ambiente*

* Apenas UMA forma de configurar variáveis (justamente… variáveis de ambiente)
* Apenas UMA nomenclatura de para essa variáveis

*Módulo de configuração*

* Flexibilidade para essas duas coisas
* Possibilidade de sobrescrever temporariamente as configurações de ambiente
* Possibilidade de validar e converter tipos caso necessário

**Comment by FelipeSBarros** [comment]
Sobre o módulo `settings`:
Vi que CASO não exista nenhuma variável `FOGOCRUZADO_EMAIL` ou `FOGOCRUZADO_PASSWORD` no `.env` (ou settings.ini), é retornado um erro `UndefinedValueError` do python-decouple. Isso acontece, mesmo que seja passado o valor por parâmetro da função.
Ou seja, o usuário seria obrigado a criar um .env e adicionar tais contstantes no arquivo.
Entendo que é por isso que vc menciona:

> Apenas UMA forma de configurar variáveis (justamente… variáveis de ambiente)

Seria o caso de termos um `settings.ini` no projeto, com essas variáveis vazias de forma que o usuário possa alterar-las com seus dados ou, apenas ignorar adicionando tais variáveis como parâmetros da função, evitando assim o `UndefinedValueError`?
Depois de uma hora redatando, testando e lendo documentção, parece que essa é a solução óbiva...
Seria essa a sua proposta?

**Comment by cuducos** [comment]
> Seria o caso de termos um `settings.ini` no projeto, com essas variáveis vazias de forma que o usuário possa alterar-las com seus dados ou, apenas ignorar adicionando tais variáveis como parâmetros da função, evitando assim o `UndefinedValueError`?

Não. Você está confundindo aplicação com configuração. A solução para esse problema que você aponta é (lista não exclusiva):

- melhorar documentação (incluindo colocar `.env.sample` como exemplo, mas não se limitando a isso
- personalizar a mensagem de erro (já viu mensagem de erro do Elm ou do Rust? Pesquise sobre isso que é interessante)
- utilizar ferramentas que ajudem a pessoa a configurar a aplicação (por exemplo, [`createnv`](https://github.com/cuducos/createnv))

**Reply by FelipeSBarros** [reply]
Entendo.
"Pensando alto":
Estariamos definindo
> Apenas UMA forma de configurar variáveis (justamente… variáveis de ambiente)

E essa forma seria pelo `.env`ou `settings.ini`. OK. 
Mas e se a pessoa quiser usar apenas passando essas informações de e-mail e password como parâmetros da função?
Que, por sinal, relendo as mensagens anteriores, teria relação com a documentação de casos específicos, como quando vc diz:
> Aí a gente documenta os casos personalizados… por exemplo, senão for de configurar FOGOCRUZADO_EMAIL e FOGOCRUZADO_PASSWORD:

Entendo que o que vc estava propondo é: se não for configurar o FOGOCRUZADO_EMAIL e FOGOCRUZADO_PASSWORD no `.env` ou `settings.ini`, poderia passar tais infos como parâmetro da classe. Contudo, ao não existir essas constantes no `.env`, o python-decouple retornaria `UndefinedValueError`, a não ser que já definammos um default vazio...:
` self.email =  kwargs.get('email', config('FOGOCRUZADO_EMAIL', default=''))`

Posso e devo estar confundindo tudo. Entendo que como boas prácticas o certo seria separar tais variáveis, mas como aprendi a programar em R, acho _normal_ alguem querer "apenas" instalar o pacote e passar as credenciais por parâmetro sem ter que se preocupar com `.env`...

**Comment by cuducos** [comment]
Tem suas confusões aí : )

1. Você não considerou a possibilidade de capturar o `UndefinedValueError`
2. Você apenas considerou usar o valor informado na inicialização da classe _caso_ o `config` do `decouple` falhe, mas não considerou o contrário

Ou seja, por um lado, podemos fazer um erro super explicativo para quem for utilizar:

```python
try:
    self.email = config('EMAIL_MAGICO')
except UndefinedValueError:
    message =
        "O Fogo Cruzado precisa estar autenticado para funcionar, no entanto "
        "não encontramos a condiguração com o email a utilizar para essa "
        "autenticação. Existem algumas formar de configurar esse email:\n"
        "- criando uma variável de ambiente chamada EMAIL_MAGICO\n"
        "- incluindo uma linha EMAIL_MAGICO=… no arquivo `.env`\n"
        "- incluindo a variável EMAIL_MAGICO no arquivo `settings.ini`\n"
    )
    raise Exception(message)
```

Mas o que eu acho melhor é a inicialização da classe sobrescrevber a chamado o `config`. Como sugeri no exemplo anterior:


```python
    self.email =  kwargs.get('email', config('FOGOCRUZADO_EMAIL'))
```

Nesse caso, _primeiro_ a classe tenta ler os argumentos de inicialização, _depois_ chama o `config`. Ou seja, esse exemplo que você escreveu com o `default` vazio é desnecessário. No máximo combinamos as duas “confusões” para poder incluir a possibilidade de inicilização da classe como parte do erro:

```python
try:
    self.email = kwargs.get('EMAIL_MAGICO', config('EMAIL_MAGICO')) # alterei aqui
except UndefinedValueError:
    message =
        "O Fogo Cruzado precisa estar autenticado para funcionar, no entanto "
        "não encontramos a condiguração com o email a utilizar para essa "
        "autenticação. Existem algumas formar de configurar esse email:\n"
        "- criando uma variável de ambiente chamada EMAIL_MAGICO\n"
        "- incluindo uma linha EMAIL_MAGICO=… no arquivo `.env`\n"
        "- incluindo a variável EMAIL_MAGICO no arquivo `settings.ini`\n"
        "- incializando essa classe com o valor do email, por exemplo Settings(email='…')" # adicionei essa linha
    )
    raise Exception(message)
```

**Reply by FelipeSBarros** [reply]
> Você não considerou a possibilidade de capturar o UndefinedValueError

Sim :). Cheguei até a escrever o código. Só não cheguei a escrever uma mensagem nao boa quanto a sua. rs
Enfim, nessa de escrever o código, lembrei que havia a possibilidade de definir um valor default, que não levantaria o erro e permitiria a pessoa a usar as informaç~çoes por parâmetro da função... daí fui acumulando dúvidas e pensamentos... vou deixá-los um pouco de lado pois acho o que você disse em seguida é um ponto importante:

> Mas o que eu acho melhor é a inicialização da classe sobrescrevber a chamado o config. Como sugeri no exemplo anterior:

Perfeito! Contudo, ele só irá sobreescrever caso já exista um `.env` ou `settings.ini` com tais constantes. Caso contrário, o erro `UndefinedValueError` será levantado.
Ou seja, sua última linha na mensaje de erro não funcionaria:
>  "- incializando essa classe com o valor do email, por exemplo Settings(email='…')" # adicionei essa linha

Mas estou entendendo que é isso, mesmo. A ideia é lenvantar o erro e assim garantimos que exista:

> Apenas UMA forma de configurar variáveis (justamente… variáveis de ambiente)

Neste caso só removeria a ultima linha da sua mensagem...

Sim, isso já estava claro. Mas justamente por ter visto a possibilidade de fazer um _acochambramento_ de forma a permitir que usem a função sem ter `.env` e `settings.ini`, meu lado criativou entrou em parafuso.

Seguirei nessa linha e usarei sua mensajem.

Aliás, devo traduzir a mensagem?

**Reply by FelipeSBarros** [reply]
Fiz uma pequena alteração na sua mensagem, veja o que acha:
```
message = (
            "Para poder autenticar o acesso aos dados do projeto Fogo Cruzado é preciso"
            "configurar o email e senha cadastrados na API do projeto."
            "Existem algumas formas de configurar ambas informações:\n"
            "- criando as variáveis de ambiente chamadas FOGOCRUZADO_EMAIL e FOGOCRUZADO_PASSWORD\n"
            "- incluindo uma linha FOGOCRUZADO_EMAIL=... e FOGOCRUZADO_PASSWORD=... no arquivo `.env`\n"
            "- incluindo a variável FOGOCRUZADO_EMAIL=... e FOGOCRUZADO_PASSWORD=... no arquivo `settings.ini`\n"
        )
```
